static/js/stream_user_group_access_data.js

```diff
@@ -0,0 +1,91 @@
+import _ from "lodash";
+// This stores stream_user_group_access objects mapped according to their id.
+let access_objs_by_id_map;
+
+// This stores all the stream_user_group_access objects of a stream as a array
+// with key as stream_id.
+let access_objs_by_stream_map;
+export function init() {
+    access_objs_by_id_map = new Map();
+    access_objs_by_stream_map = new Map();
+}
+// init function helps to clear data in tests.
+init();
+
+const stream_user_group_access_fields = ["id", "stream_id", "group_id"];
+
+export function add_access_obj(access_obj) {
+    const cleaned_obj = _.pick(access_obj, stream_user_group_access_fields);
+    access_objs_by_id_map.set(cleaned_obj.id, cleaned_obj);
+    add_to_stream_map(access_obj);
+}
+
+export function add_to_stream_map(access_obj) {
+    // This should be called from add_access_obj only so that data in both maps is
+    // always in sync.
+
+    // Adds new stream user_group_access obj in array of a stream only if it is
+    // is already not present
+    const cleaned_obj = _.pick(access_obj, stream_user_group_access_fields);
+    if (access_objs_by_stream_map.has(cleaned_obj.stream_id)) {
+        if (!is_user_group_allowed(cleaned_obj.group_id, cleaned_obj.stream_id)) {
+            let allowed_access_objs = access_objs_by_stream_map.get(access_obj.stream_id);
+            allowed_access_objs = [...allowed_access_objs, cleaned_obj];
+            access_objs_by_stream_map.set(cleaned_obj.stream_id, allowed_access_objs);
+        } else {
+            return;
+        }
```

Comment:
Why do we need the `else` block here?

Reply:
I'll remove the second `if` and add that condition as `&&` to upper `if`. That else was used to silent `eslint` error as I wanted second if to be separate but it makes more sense with the upper one.